### ISSUE TITLE
"FreeRDP initialization may fail: The current users home directory ("/usr/sbin") is not writeable but FreeRDP generally requires a writeable home directory ...." messages

### ISSUE THREAD

**Issue by ganzgustav22** [itiligent/Guacamole-Installer#67]
Lots of "FreeRDP initialization may fail: The current users home directory ("/usr/sbin") is not writeable but FreeRDP generally requires a writeable home directory ...." messages in the logs.

Is that an issue? Home directory "/usr/sbin" sounds kind of wrong, I guess that is just where the binary is?

**Comment by itiligent** [comment]
You've not taken the time to explain any background so I can only speculate. Its likely that your permissions structure is somehow broken.

**Comment by ganzgustav22** [comment]
It's a freshly installed Debian 12 bookworm. No changes to the system except I added a line in /etc/sudoers that allows the user I ran your setup script with to use sudo without password (I think your setup script assumes it's the same way under Debian as it is under Ubuntu, where any user can just run "sudo <elevated command>").

Checking with ps aux shows guacd runs under user root, and tomcat runs under user tomcat.

/usr/sbin is writeable only to root (seems fine to me).

There is one user in /etc/passwd which has /usr/sbin as home dir, it's the user "daemon".

Checking with ps aux while being connected via RDP to a windows machine though shows nothing running under user "daemon".

**Comment by itiligent** [comment]
Ok that explains it. 

You really should follow the script and not install as root -  there are some very good reasons for installing as a user, and only elevating some install tasks as sudo. 

If you install everything as root, you break a number of quite delicate security and permissions settings -not only with  freerdp (which needs a writable home dir for certificates and profile stuff, the reason for the log errors) but also with how the guac binary needs to be installed with permisions that dont allow interactive logon (so if guac is compromised, no root!) and also some other database things... Its a complex install to get right, and installing everything as root is never how to build a server and in this case wont allow all these things work correctly as intended.

Theres a lot of junk "guides" out there on installing guac, they are pretty much all wrong or incomplete. This script represents many 100s of hours of work over serveal years, and its built on the old guac-install script which goes back another few more years... why devaite?

**Comment by ganzgustav22** [comment]
I did not install as root.

I used an unpriveledged user that was added to /etc/sudoers to run the script.

**Comment by itiligent** [comment]
Another thought and more background....

Editing things so a secure appliance does not need a root password (yet installs key services as root )is not a use case that you can expect not to have issues.

See insall script lines below...
guacd run under 'daemon', not root! Freerdp also needs to run as 'daemon' too in order to have permissions to talk to guacd. This way if either guac or freerdp are compromised, that cannot be interatively used to elevate access as daemon is a very restricted accout that blocks interactive logon. Installing without sudo re-auth (a just in time re-auth)  breaks this.
 
```
# Setup freerdp profile permissions for storing certificates
mkdir -p /usr/sbin/.config/freerdp
chown daemon:daemon /usr/sbin/.config/freerdp

# Setup correct permissions for history recorded storage feature
mkdir -p /var/guacamole
chown daemon:daemon /var/guacamole
```
You could change the above to root, but that would be about as silly one could break it .... i.e. a guac server running as root and no need for password re-authentication from sodoers file.

Just install debian with sudo, create a user account, run the script and in 3 mins you will be 100 running.

**Comment by ganzgustav22** [comment]
Again, I did not edit anything.

**Comment by ganzgustav22** [comment]
Checked the permissions of /usr/sbin/.config/freerdp and /var/guacamole, both daemon:daemon.

**Comment by ganzgustav22** [comment]
> Editing things so a secure appliance does not need a root password (yet installs key services as root )is not a use case that you can expect not to have issues.

This is standard ubuntu behavior. Atleast it was when I last used ubuntu, when it was ubuntu 16.04.
I did this only to make your script run, otherwise it would ask for the sudo password over and over again.
Ofcourse I removed it later.

**Comment by itiligent** [comment]
The issue is not the script or any instructions, it is related to some basic Linux admin concepts. Lets step back a bit..

1. You added a user to the sudo group, under su - 
2.  And then stepped back to that user's **same shell** with the expectation the user should no have have full sudo ritghts.

Q: What is missing?
A: Because you are still in the old shell that was opened without the user having sudo rights,thats the current environement. Groupd rights are not updated dynamically in the terminal. Under Linux terminal you need to:
1. logout and login again
or
2. Start a new shell
or
3. run `newgrp sudo` to refresh current shell permissions
4. One final way is simply to install sudo at debian install. Sudo is added at install only if you choose not to run as root in the initial options and create a named 1st user. This user has sudo privs from the get go.

**Comment by ganzgustav22** [comment]
That worked. As a debian user that uses "su -" to become root ever since, I'm certainly no sudo expert.

Maybe consider adding the above info to the message that suggests to add the user to the sudo group. Apart from that, I still don't think that adding a user to the sudo group works with the sudo command. Atleast not with a standard debian install, the user is not in the sudo group and there is nothing else configured in /etc/sudoers. The debian installer asks to create a user during installation and that was the user I used for starting the script.. I am not aware of any initial options to choose to not run as root during installation or install sudo by default. But maybe I've just been overlooking that option for the last 20 years.



Regarding the initial issue, the log message in the title of this Github issue: 

The log message still occurs after the fresh installation. guacd now runs with user daemon (it was indeed my mistake that it ran as root before, I changed the user in the systemd unit file without remembering that I did that, I just noticed now that I compared both file on the old and newly installed VM).

However, after looking more closely at it, it's quite clear why that message appears. The user "daemon" has home directory /usr/sbin, and /usr/sbin is not writeable for user "daemon", so logically, that message appears.

And I'm pretty sure it appears for everybody on every distro, I don't think any distro has /usr/sbin writeable for the daemon user. This message has nothing to do with my lack of sudo knowledge.

Apart from that, in general, using /usr/sbin as a home dir for storing data doesn't seem like a good idea, that's not what that directory is intended for. I guess you probably know that by now, maybe that is the reason why you deleted your message in this thread that showed the places in your script where you create and chown .config/freeerdp directories under /usr/sbin. Here (https://github.com/itiligent/Guacamole-Installer/issues/67#issuecomment-2444053276) I quoted that message, but your original message is gone.

Also, what you wrote in that deleted message about guacd and freerdp needing the same user/group to run is wrong. Actually, there is no distinct freerdp daemon or binary, guacd just uses the freerdp client library.


tl;dr: To get rid of this log message and fix the bad idea of storing freerdp data under /usr/sbin, the following works:

- Create a system user guacd with home directory /var/lib/guacd (or whatever other suitable place)
- Re-Create the directory structure .config/freerdp/... that was under /usr/sbin before under /var/lib/guacd
- chown everything to guacd
- changed guacd systemd unit file to use user guacd user instead of daemon user

Voila, log message gone, data files under a more sane location than /usr/sbin

**Comment by itiligent** [comment]
Here is more context to the freerdp home dir issue  -  Its bascially a legacy workaround for freerdp quirks. Without it you might find not everthing works.  

After guac 1.2, freedrp got broken coz now guacd runs under daemon, but the contradiction is that freerdp needs a home dir to write certs. The /usr/sbin/.config/freerdp dir should be a writeable workaround. Importantly, under the hood freerdp needs the same privs as guacd to allow screen recording, so this is the catch22 that the home dir workaround solves.

If you are still seeing errors you've got something with the freerdp dir not writeable somehow. If systemd file is not running exactly as intended this might explain your logs? I've never seen the errors you are seeing.

Some deeper discussion on this now seems ro have broken links as is a few yers old now. See here for many others with same solution... 
https://lists.apache.org/thread/x1ljjckn5dqf50zj7p79wpn99299cxhh
https://issues.apache.org/jira/browse/GUACAMOLE-1249 (note this is for 1.2 pre deamon change, so moot for 1,3 onwards)

If you've found a better way, of course please submit a PR coz that would be very helpful. This is free code after all!

Lastly I deleted earier posts only after more context was provided from you. This is to keep things on point and not mislead future readers, not out of incorrectness. I somtimes change the issue subjects too, just to help make things more searchable for others who may bump into similar problems.

**Comment by itiligent** [comment]
oh, and during debian install you must not enable root account by default... this installs sudo and gives you a default ubuntu like default. `sudo passwd root` will enable root after install,  allowing both su elevaton and sudo

**Comment by itiligent** [comment]
and this.. https://github.com/MysticRyuujin/guac-install/issues/197

**Comment by itiligent** [comment]
I'm stil trying to untangle what you've got. It is possible that the `/usr/sbin/.config/freerdp` directoy is not created due to the order things have been done?

**Comment by itiligent** [comment]
found this for you too. skip to  1:45 https://www.youtube.com/watch?v=z0Cs8c84r4k

**Comment by itiligent** [comment]
Just tidying up this thread for future readers - Hence a repost with a bit more succintness.

- FYI I deleted earlier posts once I learned about your addtional custom local changes - this helped me immediately see this was as Linux administraion issue, so nothing I originally wrote about messing with a root install was relevant to future readers. Its always better to put effort into providing good detail up front, rather than into discrediting free help... yes? 

- I have taken you feedback and updated the script's sudo warning to include a reminder to re-login.

-  Be advised there is deliberate method in the use of  /usr/sbin/.config/freerdp as a quasi daemon home dir.  After guacammole 1.2 guacd runs by default under 'daemon' (no home dir), but freerdp needs a place to store certificates. For freerdp screen recording to work, it expects the same user calling it to have a home dir to pre present in that user's path...therefore the permissions in the script (AFAIK) are absolutely required as written. Of course you can change things to run guacd as a regualr user too, but this compromises the higher level of security offered by running guacd as its default 'daemon' which is far more restricted and does not allow interactive shell login. 
If you've found a better way that maintains the security model and has freerdp 100% functional, of course please submit a PR coz that would be very helpful.  
 
Some old discussion on this issue in apache archives seems to be broken links by now, but see here for more background 
https://lists.apache.org/thread/x1ljjckn5dqf50zj7p79wpn99299cxhh
https://github.com/MysticRyuujin/guac-install/issues/196
https://issues.apache.org/jira/browse/GUACAMOLE-1249 (note this is for 1.2 pre deamon change, so moot for 1,3 onwards)

This link might actually be what you want, but its not added yet. Not the caution from the devs about breaking the exisitng model by running as another user:  https://lists.apache.org/list?dev@guacamole.apache.org:2024-9:daemon

If you've found a better way that maintains the security model and has freerdp 100% functional, of course please submit a PR coz that would be very helpful.

**Comment by itiligent** [comment]
FYI I made a quick change to allow the new option of setting a specific service account.  

Take a look in the setup script variables to set it to something other than daemon  - the script will do the rest. 

I cant speak to the exact security impact running this way makes, but I've locked down the from login. I might lock it out of ssh access too.
